### PR TITLE
Fix duplicated column definition in the gpbackup metadata

### PR DESCRIPTION
**Pull request by whitehawk**
Fix duplicated column definition in the gpbackup metadata

Problem description:
If a column in a table had privileges granted for two (or more) users, gpbackup
created metadata file, where this column was duplicated.

Cause:
The SQL query, that is used to retrieve information about all table columns in 
the function 'GetColumnDefinitions', had the 'unnest' function applied to the 
array of column-level access privileges from the 'pg_catalog.pg_attribute'.
Because of the 'unnest', if there were several aclitems for a column, the result
of the query had several rows returned for this column. But gpbackup treated all
these rows as results for separate columns, causing duplication in the metadata.

Fix:
'unnest' is removed in the query, 'attacl' is returned as is. Thus, for each 
column, there is only one row in the result of the query. Now, the field 
'Privileges' in the 'ColumnDefinition' struct contains all column-level access 
privileges in a slice of strings ('pq.StringArray'). And later, in 
'getColumnACL', this slice, where each string contains a separate aclitem, is 
iterated and each item is parsed as before. Plus, the 'Kind' field is removed
from the 'ColumnDefinition' struct.

### CODE REVIEW

**Comment by RekGRpth** [comment]
this diff
```diff
diff --git a/end_to_end/end_to_end_suite_test.go b/end_to_end/end_to_end_suite_test.go
index 334e979e..46b189ab 100644
--- a/end_to_end/end_to_end_suite_test.go
+++ b/end_to_end/end_to_end_suite_test.go
@@ -2013,11 +2013,11 @@ LANGUAGE plpgsql NO SQL;`)
 					Skip("This test is not needed for old backup versions")
 				}
 				testhelper.AssertQueryRuns(backupConn, "CREATE TABLE public.t1 (a int) DISTRIBUTED BY (a);")
-				testhelper.AssertQueryRuns(backupConn, "CREATE USER user1;")
+				testhelper.AssertQueryRuns(backupConn, "CREATE USER \"user,1\";")
 				testhelper.AssertQueryRuns(backupConn, "CREATE USER user2;")
-				testhelper.AssertQueryRuns(backupConn, "GRANT SELECT (a) on public.t1 to user1;")
+				testhelper.AssertQueryRuns(backupConn, "GRANT SELECT (a) on public.t1 to \"user,1\";")
 				testhelper.AssertQueryRuns(backupConn, "GRANT SELECT (a) on public.t1 to user2;")
-				defer testhelper.AssertQueryRuns(backupConn, "DROP USER user1;")
+				defer testhelper.AssertQueryRuns(backupConn, "DROP USER \"user,1\";")
 				defer testhelper.AssertQueryRuns(backupConn, "DROP USER user2;")
 				defer testhelper.AssertQueryRuns(backupConn, "DROP TABLE public.t1;")
 
@@ -2025,7 +2025,7 @@ LANGUAGE plpgsql NO SQL;`)
 
 				contents := string(getMetdataFileContents(backupDir, timestamp, "metadata.sql"))
 				Expect(contents).To(ContainSubstring("CREATE TABLE public.t1 (\n\ta integer\n) DISTRIBUTED BY (a);"))
-				Expect(contents).To(ContainSubstring("GRANT SELECT (a) ON TABLE public.t1 TO user1;"))
+				Expect(contents).To(ContainSubstring("GRANT SELECT (a) ON TABLE public.t1 TO \"user,1\";"))
 				Expect(contents).To(ContainSubstring("GRANT SELECT (a) ON TABLE public.t1 TO user2;"))
 			})
 		})
```
breaks patch
```sql
...
      GRANT SELECT (a) ON TABLE public.t1 TO 1";
      GRANT SELECT (a) ON TABLE public.t1 TO user2;
...
```
try to use pq.StringArray instead sql.NullString for Privileges field in ColumnDefinition struct

**Comment by whitehawk** [comment]
> this diff
> 
> ```diff
> diff --git a/end_to_end/end_to_end_suite_test.go b/end_to_end/end_to_end_suite_test.go
> index 334e979e..46b189ab 100644
> --- a/end_to_end/end_to_end_suite_test.go
> +++ b/end_to_end/end_to_end_suite_test.go
> @@ -2013,11 +2013,11 @@ LANGUAGE plpgsql NO SQL;`)
>  					Skip("This test is not needed for old backup versions")
>  				}
>  				testhelper.AssertQueryRuns(backupConn, "CREATE TABLE public.t1 (a int) DISTRIBUTED BY (a);")
> -				testhelper.AssertQueryRuns(backupConn, "CREATE USER user1;")
> +				testhelper.AssertQueryRuns(backupConn, "CREATE USER \"user,1\";")
>  				testhelper.AssertQueryRuns(backupConn, "CREATE USER user2;")
> -				testhelper.AssertQueryRuns(backupConn, "GRANT SELECT (a) on public.t1 to user1;")
> +				testhelper.AssertQueryRuns(backupConn, "GRANT SELECT (a) on public.t1 to \"user,1\";")
>  				testhelper.AssertQueryRuns(backupConn, "GRANT SELECT (a) on public.t1 to user2;")
> -				defer testhelper.AssertQueryRuns(backupConn, "DROP USER user1;")
> +				defer testhelper.AssertQueryRuns(backupConn, "DROP USER \"user,1\";")
>  				defer testhelper.AssertQueryRuns(backupConn, "DROP USER user2;")
>  				defer testhelper.AssertQueryRuns(backupConn, "DROP TABLE public.t1;")
>  
> @@ -2025,7 +2025,7 @@ LANGUAGE plpgsql NO SQL;`)
>  
>  				contents := string(getMetdataFileContents(backupDir, timestamp, "metadata.sql"))
>  				Expect(contents).To(ContainSubstring("CREATE TABLE public.t1 (\n\ta integer\n) DISTRIBUTED BY (a);"))
> -				Expect(contents).To(ContainSubstring("GRANT SELECT (a) ON TABLE public.t1 TO user1;"))
> +				Expect(contents).To(ContainSubstring("GRANT SELECT (a) ON TABLE public.t1 TO \"user,1\";"))
>  				Expect(contents).To(ContainSubstring("GRANT SELECT (a) ON TABLE public.t1 TO user2;"))
>  			})
>  		})
> ```
> 
> breaks patch
> 
> ```sql
> ...
>       GRANT SELECT (a) ON TABLE public.t1 TO 1";
>       GRANT SELECT (a) ON TABLE public.t1 TO user2;
> ...
> ```
> 
> try to use pq.StringArray instead sql.NullString for Privileges field in ColumnDefinition struct

fixed

File: backup/queries_table_defs.go [hunk 1]
```diff
@@ -302,10 +303,7 @@ func GetColumnDefinitions(connectionPool *dbconn.DBConn) map[uint32][]ColumnDefi
 		CASE WHEN a.attstorage != t.typstorage THEN a.attstorage ELSE '' END AS storagetype,
 		coalesce('('||pg_catalog.pg_get_expr(ad.adbin, ad.adrelid)||')', '') AS defaultval,
 		coalesce(d.description, '') AS comment,
-		CASE
-			WHEN a.attacl IS NULL THEN NULL
-			WHEN array_upper(a.attacl, 1) = 0 THEN a.attacl[0]
-			ELSE unnest(a.attacl) END AS privileges,
+		a.attacl AS privileges,
 		CASE
```
Comment (RekGRpth) [inline backup/queries_table_defs.go:307]: I think, we may remove this and rewrite logic related to kind Empty

Comment (RekGRpth) [reply backup/queries_table_defs.go]: Check other places with `Empty` word in code and test, for example
https://github.com/arenadata/gpbackup/blob/e75a86aa7c785f35f6f24cdc903828b8b481bba8/backup/predata_acl.go#L159
https://github.com/arenadata/gpbackup/blob/e75a86aa7c785f35f6f24cdc903828b8b481bba8/backup/predata_acl.go#L191
https://github.com/arenadata/gpbackup/blob/e75a86aa7c785f35f6f24cdc903828b8b481bba8/backup/predata_acl.go#L571
https://github.com/arenadata/gpbackup/blob/e75a86aa7c785f35f6f24cdc903828b8b481bba8/backup/predata_acl_test.go#L313
https://github.com/arenadata/gpbackup/blob/e75a86aa7c785f35f6f24cdc903828b8b481bba8/backup/predata_acl_test.go#L430
https://github.com/arenadata/gpbackup/blob/e75a86aa7c785f35f6f24cdc903828b8b481bba8/backup/queries_acl.go#L168
https://github.com/arenadata/gpbackup/blob/e75a86aa7c785f35f6f24cdc903828b8b481bba8/backup/queries_acl.go#L347
https://github.com/arenadata/gpbackup/blob/e75a86aa7c785f35f6f24cdc903828b8b481bba8/backup/queries_table_defs.go#L311
https://github.com/arenadata/gpbackup/blob/e75a86aa7c785f35f6f24cdc903828b8b481bba8/backup/queries_table_defs.go#L359
https://github.com/arenadata/gpbackup/blob/e75a86aa7c785f35f6f24cdc903828b8b481bba8/backup/queries_acl_test.go#L103
https://github.com/arenadata/gpbackup/blob/e75a86aa7c785f35f6f24cdc903828b8b481bba8/backup/predata_acl_test.go#L358
https://github.com/arenadata/gpbackup/blob/e75a86aa7c785f35f6f24cdc903828b8b481bba8/backup/predata_acl_test.go#L474
and so on

Comment (whitehawk) [reply backup/queries_table_defs.go]: The other places (that are yet not changed) are related to other structs like `MetadataQueryStruct` , not to `ColumnDefinition`. So, changing other places will be refactoring, that we do not want to mix with the bug fix.

File: backup/predata_relations_tables_test.go [hunk 2]
```diff
@@ -653,8 +654,8 @@ COMMENT ON COLUMN public.tablename.i IS 'This is a column comment.';
 COMMENT ON COLUMN public.tablename.j IS 'This is another column comment.';`)
 		})
 		It("prints a GRANT statement on a table column", func() {
-			privilegesColumnOne := backup.ColumnDefinition{Oid: 0, Num: 1, Name: "i", Type: "integer", StatTarget: -1, Privileges: sql.NullString{String: "testrole=r/testrole", Valid: true}}
-			privilegesColumnTwo := backup.ColumnDefinition{Oid: 1, Num: 2, Name: "j", Type: "character varying(20)", StatTarget: -1, Privileges: sql.NullString{String: "testrole2=arwx/testrole2", Valid: true}}
+			privilegesColumnOne := backup.ColumnDefinition{Oid: 0, Num: 1, Name: "i", Type: "integer", StatTarget: -1, Privileges: pq.StringArray{"testrole=r/testrole"}}
```
Comment (RekGRpth) [inline backup/predata_relations_tables_test.go:657]: may be for some of these columns use 2 acls in array for test?

File: integration/predata_relations_create_test.go [hunk 3]
```diff
@@ -422,7 +423,7 @@ SET SUBPARTITION TEMPLATE ` + `
 		})
 		It("prints column level privileges", func() {
 			testutils.SkipIfBefore6(connectionPool)
-			privilegesColumnOne := backup.ColumnDefinition{Oid: 0, Num: 1, Name: "i", Type: "integer", StatTarget: -1, Privileges: sql.NullString{String: "testrole=r/testrole", Valid: true}}
+			privilegesColumnOne := backup.ColumnDefinition{Oid: 0, Num: 1, Name: "i", Type: "integer", StatTarget: -1, Privileges: pq.StringArray{"testrole=r/testrole"}}
```
Comment (RekGRpth) [inline integration/predata_relations_create_test.go:426]: may be use 2 acls in array for test?

Comment (whitehawk) [reply integration/predata_relations_create_test.go]: ok

File: backup/predata_relations_tables_test.go (hunk 2 above)
Comment (whitehawk) [reply backup/predata_relations_tables_test.go]: ok

**Review by RekGRpth — APPROVED** [review]

**Review by bimboterminator1 — APPROVED** [review]

**Pull request merged**